File: render_chart.py
# /// script
# requires-python = ">=3.10"
# dependencies = [
#   "matplotlib>=3.7",
#   "numpy>=1.24",
# ]
# ///
"""
render_chart.py — Beautiful Chart Renderer
Usage: uv run render_chart.py chart_config.json [output.png]

If output path is given  → saves PNG file to that path
If output path is absent → prints base64 JSON block to stdout so Claude Code
                           renders the image inline in the chat automatically
"""
import sys, json, io, base64
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
from matplotlib.lines import Line2D
import matplotlib.patches as mpatches
import numpy as np

# ── Design tokens ─────────────────────────────────────────────────────────────
PALETTE = {
    'blue':   '#185FA5',
    'red':    '#E24B4A',
    'teal':   '#1D9E75',
    'amber':  '#BA7517',
    'purple': '#534AB7',
    'gray':   '#888780',
}
PALETTE_ORDER = list(PALETTE.keys())

FONT        = 'Liberation Sans'
COLOR_TEXT  = '#111111'
COLOR_MUTED = '#999999'
COLOR_FAINT = '#bbbbbb'
COLOR_GRID  = '#e8e8e8'
COLOR_SPINE = '#e0e0e0'

plt.rcParams.update({
    'font.family':        FONT,
    'font.size':          11,
    'axes.spines.top':    False,
    'axes.spines.right':  False,
    'axes.spines.left':   False,
    'axes.spines.bottom': True,
    'axes.grid':          True,
    'axes.axisbelow':     True,
    'grid.alpha':         1.0,
    'grid.color':         COLOR_GRID,
    'grid.linewidth':     0.6,
    'xtick.bottom':       False,
    'ytick.left':         False,
    'xtick.major.pad':    6,
    'ytick.major.pad':    6,
})


def color_for(name, idx=0):
    if name in PALETTE:
        return PALETTE[name]
    return PALETTE[PALETTE_ORDER[idx % len(PALETTE_ORDER)]]


def base_ax(fig, ax):
    fig.patch.set_alpha(0)
    ax.set_facecolor('none')
    ax.spines['bottom'].set_color(COLOR_SPINE)
    ax.spines['bottom'].set_linewidth(0.7)
    ax.tick_params(colors=COLOR_MUTED, labelsize=10.5)


def add_labels(fig, title, subtitle, source):
    y = 0.98
    if title:
        fig.text(0.0, y, title, ha='left', va='top', transform=fig.transFigure,
                 fontsize=15, fontweight='medium', color=COLOR_TEXT, font=FONT)
        y -= 0.075
    if subtitle:
        fig.text(0.0, y, subtitle, ha='left', va='top', transform=fig.transFigure,
                 fontsize=11, color=COLOR_MUTED, font=FONT)
    if source:
        fig.text(0.0, 0.0, source, ha='left', va='bottom', transform=fig.transFigure,
                 fontsize=9, color=COLOR_FAINT, font=FONT)


def fmt_axis(ax, y_prefix, y_suffix, ymin, ymax, x_labels=None, horizontal=False):
    fmt = mticker.FuncFormatter(lambda v, _: f'{y_prefix}{v:g}{y_suffix}')
    if horizontal:
        ax.xaxis.set_major_formatter(fmt)
    else:
        ax.yaxis.set_major_formatter(fmt)
        if x_labels is not None:
            ax.set_xticks(range(len(x_labels)))
            ax.set_xticklabels(x_labels, color=COLOR_MUTED)
    if ymin is not None:
        if horizontal: ax.set_xlim(left=ymin)
        else:          ax.set_ylim(bottom=ymin)
    if ymax is not None:
        if horizontal: ax.set_xlim(right=ymax)
        else:          ax.set_ylim(top=ymax)


def line_legend(ax, datasets):
    handles = [
        Line2D([0],[0], color=color_for(ds.get('color',''), i),
               linewidth=2, marker='o', markersize=5,
               label=ds.get('label', f'Series {i+1}'))
        for i, ds in enumerate(datasets)
    ]
    ax.legend(handles=handles, frameon=False, labelcolor=COLOR_MUTED,
              fontsize=10, loc='upper right',
              handlelength=1.6, borderpad=0, labelspacing=0.35)


def patch_legend(ax_or_fig, handles, loc='upper right', anchor=None):
    kw = dict(frameon=False, labelcolor=COLOR_MUTED, fontsize=10,
              handlelength=1.0, borderpad=0, labelspacing=0.4)
    if anchor:
        ax_or_fig.legend(handles=handles, loc=loc, bbox_to_anchor=anchor, **kw)
    else:
        ax_or_fig.legend(handles=handles, loc=loc, **kw)
    leg = ax_or_fig.get_legend() if hasattr(ax_or_fig, 'get_legend') else None
    if leg:
        for patch in leg.get_patches():
            patch.set_height(9); patch.set_width(9)


def draw_line(fig, ax, labels, datasets):
    xi = range(len(labels))
    for i, ds in enumerate(datasets):
        c = color_for(ds.get('color',''), i)
        ax.plot(xi, ds['data'], color=c, linewidth=2.2,
                marker='o', markersize=5.5, markerfacecolor=c,
                markeredgewidth=0, zorder=3, solid_capstyle='round')
        if ds.get('fill', False):
            base = min(ds['data']) - (max(ds['data']) - min(ds['data'])) * 0.1
            ax.fill_between(xi, ds['data'], base, alpha=0.06, color=c)
    line_legend(ax, datasets)


def draw_bar(fig, ax, labels, datasets):
    n = len(datasets)
    x = np.arange(len(labels))
    w = 0.65 / max(n, 1)
    handles = []
    for i, ds in enumerate(datasets):
        c = color_for(ds.get('color',''), i)
        offset = (i - (n-1)/2) * w
        ax.bar(x + offset, ds['data'], width=w*0.88, color=c, zorder=3, linewidth=0)
        handles.append(mpatches.Patch(color=c, label=ds.get('label', f'Series {i+1}')))
    ax.set_xticks(x)
    ax.set_xticklabels(labels, color=COLOR_MUTED)
    if len(handles) > 1:
        patch_legend(ax, handles)


def draw_hbar(fig, ax, labels, datasets):
    n = len(datasets)
    y = np.arange(len(labels))
    w = 0.65 / max(n, 1)
    handles = []
    for i, ds in enumerate(datasets):
        c = color_for(ds.get('color',''), i)
        offset = (i - (n-1)/2) * w
        ax.barh(y + offset, ds['data'], height=w*0.88, color=c, zorder=3, linewidth=0)
        handles.append(mpatches.Patch(color=c, label=ds.get('label', f'Series {i+1}')))
    ax.set_yticks(y)
    ax.set_yticklabels(labels, color=COLOR_MUTED)
    ax.spines['bottom'].set_visible(False)
    ax.spines['left'].set_color(COLOR_SPINE)
    ax.spines['left'].set_linewidth(0.7)
    ax.xaxis.grid(True, color=COLOR_GRID, linewidth=0.6)
    ax.yaxis.grid(False)
    if len(handles) > 1:
        patch_legend(ax, handles)


def draw_area(fig, ax, labels, datasets):
    xi = range(len(labels))
    handles = []
    for i, ds in enumerate(datasets):
        c = color_for(ds.get('color',''), i)
        ax.fill_between(xi, ds['data'], alpha=0.18, color=c)
        ax.plot(xi, ds['data'], color=c, linewidth=2, zorder=3)
        handles.append(mpatches.Patch(color=c, label=ds.get('label', f'Series {i+1}')))
    if len(handles) > 1:
        patch_legend(ax, handles)


def draw_scatter(fig, ax, labels, datasets):
    handles = []
    for i, ds in enumerate(datasets):
        c = color_for(ds.get('color',''), i)
        pts = ds['data']
        xs = [p[0] for p in pts]
        ys = [p[1] for p in pts]
        sz = [p[2]*12 if len(p) > 2 else 55 for p in pts]
        ax.scatter(xs, ys, s=sz, color=c, alpha=0.72, zorder=3, linewidths=0)
        handles.append(mpatches.Patch(color=c, label=ds.get('label', f'Series {i+1}')))
    if len(handles) > 1:
        patch_legend(ax, handles)


def draw_donut(fig, ax, labels, datasets):
    ax.axis('off')
    ds      = datasets[0]
    data    = ds['data']
    dlabels = ds.get('labels', labels)
    cnames  = ds.get('colors', PALETTE_ORDER)
    colors  = [color_for(cnames[i] if i < len(cnames) else '', i) for i in range(len(data))]
    pie_ax  = fig.add_axes([0.08, 0.08, 0.56, 0.84])
    pie_ax.set_facecolor('none')
    pie_ax.pie(data, colors=colors, startangle=90,
               wedgeprops={'linewidth': 2, 'edgecolor': 'white', 'width': 0.52})
    handles = [mpatches.Patch(color=colors[i], label=f'{dlabels[i]}  {data[i]}')
               for i in range(len(data))]
    patch_legend(fig, handles, loc='center right', anchor=(0.98, 0.5))


# ── Main renderer ─────────────────────────────────────────────────────────────

def render(cfg):
    ctype    = cfg.get('type', 'line')
    title    = cfg.get('title', '')
    subtitle = cfg.get('subtitle', '')
    source   = cfg.get('source', '')
    labels   = cfg.get('labels', [])
    datasets = cfg.get('datasets', [])
    width    = cfg.get('width',  900) / 96
    height   = cfg.get('height', 480) / 96
    ymin     = cfg.get('yMin', None)
    ymax     = cfg.get('yMax', None)
    y_prefix = cfg.get('yPrefix', '')
    y_suffix = cfg.get('ySuffix', '')

    fig, ax = plt.subplots(figsize=(width, height), dpi=150)
    base_ax(fig, ax)

    draw = {
        'line':    draw_line,
        'bar':     draw_bar,
        'hbar':    draw_hbar,
        'area':    draw_area,
        'scatter': draw_scatter,
        'donut':   draw_donut,
    }
    draw.get(ctype, draw_line)(fig, ax, labels, datasets)

    is_hbar = (ctype == 'hbar')
    fmt_axis(ax, y_prefix, y_suffix, ymin, ymax,
             x_labels=(labels if ctype not in ('hbar','scatter','donut') else None),
             horizontal=is_hbar)

    has_header = bool(title or subtitle)
    top_pad    = 0.15 if has_header else 0.05
    bot_pad    = 0.10 if source     else 0.06
    plt.subplots_adjust(left=0.07, right=0.97, top=1-top_pad, bottom=bot_pad)
    add_labels(fig, title, subtitle, source)

    buf = io.BytesIO()
    plt.savefig(buf, transparent=True, dpi=150,
                format='png', bbox_inches='tight',
                facecolor='none', edgecolor='none')
    plt.close(fig)
    buf.seek(0)
    return buf.read()


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: uv run render_chart.py config.json [output.png]')
        sys.exit(1)

    with open(sys.argv[1]) as f:
        cfg = json.load(f)

    png_bytes = render(cfg)

    if len(sys.argv) >= 3:
        # Save to file
        outfile = sys.argv[2]
        with open(outfile, 'wb') as f:
            f.write(png_bytes)
        print(f'saved → {outfile}', file=sys.stderr)
    else:
        # Print base64 JSON block to stdout — Claude Code renders this inline
        b64 = base64.b64encode(png_bytes).decode()
        print(json.dumps({
            "type": "image",
            "source": {
                "type": "base64",
                "media_type": "image/png",
                "data": b64
            }
        }))

File: render_chart.js
#!/usr/bin/env bun
/**
 * render_chart.js — Beautiful Chart Renderer (Bun + Chart.js)
 * Usage: bunx -y beautiful-chartsjs config.json [output.png]
 *
 * If output path is given  → saves PNG file to that path
 * If output path is absent → prints base64 JSON block to stdout so Claude
 *                            renders the image inline in the chat automatically
 */

import { createCanvas } from '@napi-rs/canvas';
import { readFileSync, writeFileSync } from 'node:fs';

// ── Minimal browser polyfills — must come BEFORE the Chart.js dynamic import ─
const _mkMockCanvas = () => {
  const c = createCanvas(1, 1);
  c.style = {};
  c.addEventListener = () => { };
  c.removeEventListener = () => { };
  return c;
};

globalThis.window ??= globalThis;
globalThis.navigator ??= { userAgent: 'Bun', maxTouchPoints: 0 };
globalThis.document ??= {
  createElement: (t) => t === 'canvas' ? _mkMockCanvas() : { style: {}, addEventListener() { }, removeEventListener() { } },
  querySelector: () => null,
  querySelectorAll: () => [],
  addEventListener() { },
  removeEventListener() { },
};

// ── Chart.js loaded after polyfills so its module-init sees a mock window ───
const { Chart, registerables } = await import('chart.js');
Chart.register(...registerables);

// ── Design tokens ────────────────────────────────────────────────────────────
const PALETTE = {
  blue: '#185FA5',
  red: '#E24B4A',
  teal: '#1D9E75',
  amber: '#BA7517',
  purple: '#534AB7',
  gray: '#888780',
};
const PALETTE_ORDER = Object.keys(PALETTE);

const COLOR_TEXT = '#111111';
const COLOR_MUTED = '#999999';
const COLOR_FAINT = '#bbbbbb';
const COLOR_GRID = '#e8e8e8';
const COLOR_SPINE = '#e0e0e0';

// ── Helpers ──────────────────────────────────────────────────────────────────
const colorFor = (name, i = 0) =>
  PALETTE[name] ?? PALETTE[PALETTE_ORDER[i % PALETTE_ORDER.length]];

const rgba = (hex, a) =>
  `rgba(${parseInt(hex.slice(1, 3), 16)},${parseInt(hex.slice(3, 5), 16)},${parseInt(hex.slice(5, 7), 16)},${a})`;

// ── Custom plugin: title / subtitle / source ─────────────────────────────────
Chart.register({
  id: 'headerFooter',
  beforeDraw({ ctx, width, height, options: { plugins: p } }) {
    const hf = p?.headerFooter;
    if (!hf) return;
    const PAD = 14;
    ctx.save();
    ctx.textAlign = 'left';
    if (hf.title) {
      ctx.font = '600 18px Arial,Helvetica,sans-serif';
      ctx.fillStyle = COLOR_TEXT;
      ctx.textBaseline = 'top';
      ctx.fillText(hf.title, PAD, PAD);
    }
    if (hf.subtitle) {
      ctx.font = '13px Arial,Helvetica,sans-serif';
      ctx.fillStyle = COLOR_MUTED;
      ctx.textBaseline = 'top';
      ctx.fillText(hf.subtitle, PAD, PAD + (hf.title ? 26 : 0));
    }
    if (hf.source) {
      ctx.font = '10px Arial,Helvetica,sans-serif';
      ctx.fillStyle = COLOR_FAINT;
      ctx.textBaseline = 'bottom';
      ctx.fillText(hf.source, PAD, height - 6);
    }
    ctx.restore();
  },
});

// ── Scale factories ──────────────────────────────────────────────────────────
const valScale = (pre, suf, mn, mx) => ({
  grid: { color: COLOR_GRID, lineWidth: 0.6 },
  border: { display: false },
  ticks: { color: COLOR_MUTED, font: { size: 11 }, callback: v => `${pre}${v}${suf}` },
  ...(mn != null && { min: mn }),
  ...(mx != null && { max: mx }),
});

const catScale = () => ({
  grid: { display: false },
  border: { color: COLOR_SPINE, width: 0.7 },
  ticks: { color: COLOR_MUTED, font: { size: 11 } },
});

// ── Chart-type builders (return { type, data, typeOptions }) ─────────────────
function cfgLine({ labels, datasets, pre, suf, ymin, ymax }) {
  return {
    type: 'line',
    data: {
      labels,
      datasets: datasets.map((ds, i) => {
        const c = colorFor(ds.color, i);
        return {
          label: ds.label || `Series ${i + 1}`,
          data: ds.data,
          borderColor: c,
          backgroundColor: ds.fill ? rgba(c, 0.06) : 'transparent',
          borderWidth: 2.2,
          pointRadius: 5.5,
          pointHoverRadius: 5.5,
          pointBackgroundColor: c,
          pointBorderWidth: 0,
          fill: ds.fill ? 'origin' : false,
          tension: 0,
        };
      }),
    },
    typeOptions: { scales: { x: catScale(), y: valScale(pre, suf, ymin, ymax) } },
  };
}

function cfgBar({ labels, datasets, pre, suf, ymin, ymax }) {
  return {
    type: 'bar',
    data: {
      labels,
      datasets: datasets.map((ds, i) => ({
        label: ds.label || `Series ${i + 1}`,
        data: ds.data,
        backgroundColor: colorFor(ds.color, i),
        borderWidth: 0,
      })),
    },
    typeOptions: { scales: { x: catScale(), y: valScale(pre, suf, ymin, ymax) } },
  };
}

function cfgHBar({ labels, datasets, pre, suf, ymin, ymax }) {
  return {
    type: 'bar',
    data: {
      labels,
      datasets: datasets.map((ds, i) => ({
        label: ds.label || `Series ${i + 1}`,
        data: ds.data,
        backgroundColor: colorFor(ds.color, i),
        borderWidth: 0,
      })),
    },
    typeOptions: {
      indexAxis: 'y',
      scales: {
        x: { ...valScale(pre, suf, ymin, ymax), grid: { color: COLOR_GRID, lineWidth: 0.6 } },
        y: { ...catScale(), grid: { display: false } },
      },
    },
  };
}

function cfgArea({ labels, datasets, pre, suf, ymin, ymax }) {
  return {
    type: 'line',
    data: {
      labels,
      datasets: datasets.map((ds, i) => {
        const c = colorFor(ds.color, i);
        return {
          label: ds.label || `Series ${i + 1}`,
          data: ds.data,
          borderColor: c,
          backgroundColor: rgba(c, 0.18),
          borderWidth: 2,
          pointRadius: 0,
          fill: 'origin',
          tension: 0,
        };
      }),
    },
    typeOptions: { scales: { x: catScale(), y: valScale(pre, suf, ymin, ymax) } },
  };
}

function cfgScatter({ datasets, pre, suf, ymin, ymax }) {
  return {
    type: 'bubble',
    data: {
      datasets: datasets.map((ds, i) => {
        const c = colorFor(ds.color, i);
        return {
          label: ds.label || `Series ${i + 1}`,
          data: ds.data.map(p => ({ x: p[0], y: p[1], r: p[2] ? p[2] * 3 : 7 })),
          backgroundColor: rgba(c, 0.72),
          borderWidth: 0,
        };
      }),
    },
    typeOptions: {
      scales: {
        x: { ...catScale(), grid: { color: COLOR_GRID, lineWidth: 0.6 }, border: { display: false } },
        y: valScale(pre, suf, ymin, ymax),
      },
    },
  };
}

function cfgDonut({ datasets }) {
  const ds = datasets[0] || {};
  const cnames = ds.colors || PALETTE_ORDER;
  const colors = (ds.data || []).map((_, i) => colorFor(cnames[i] || '', i));
  return {
    type: 'doughnut',
    data: {
      labels: ds.labels || [],
      datasets: [{ data: ds.data || [], backgroundColor: colors, borderWidth: 2, borderColor: '#ffffff' }],
    },
    typeOptions: { cutout: '52%' },
  };
}

function cfgPie({ datasets }) {
  const ds = datasets[0] || {};
  const cnames = ds.colors || PALETTE_ORDER;
  const colors = (ds.data || []).map((_, i) => colorFor(cnames[i] || '', i));
  return {
    type: 'pie',
    data: {
      labels: ds.labels || [],
      datasets: [{ data: ds.data || [], backgroundColor: colors, borderWidth: 2, borderColor: '#ffffff' }],
    },
    typeOptions: {},
  };
}

function cfgPolarArea({ datasets }) {
  const ds = datasets[0] || {};
  const cnames = ds.colors || PALETTE_ORDER;
  const colors = (ds.data || []).map((_, i) => colorFor(cnames[i] || '', i));
  return {
    type: 'polarArea',
    data: {
      labels: ds.labels || [],
      datasets: [{ data: ds.data || [], backgroundColor: colors.map(c => rgba(c, 0.72)), borderWidth: 1, borderColor: '#ffffff' }],
    },
    typeOptions: {
      scales: {
        r: {
          grid: { color: COLOR_GRID, lineWidth: 0.6 },
          ticks: { color: COLOR_MUTED, font: { size: 10 }, backdropColor: 'transparent' },
        },
      },
    },
  };
}

function cfgRadar({ labels, datasets }) {
  return {
    type: 'radar',
    data: {
      labels,
      datasets: datasets.map((ds, i) => {
        const c = colorFor(ds.color, i);
        return {
          label: ds.label || `Series ${i + 1}`,
          data: ds.data,
          borderColor: c,
          backgroundColor: rgba(c, 0.18),
          borderWidth: 2,
          pointRadius: 4,
          pointBackgroundColor: c,
        };
      }),
    },
    typeOptions: {
      scales: {
        r: {
          grid: { color: COLOR_GRID, lineWidth: 0.6 },
          angleLines: { color: COLOR_GRID, lineWidth: 0.6 },
          pointLabels: { color: COLOR_MUTED, font: { size: 11 } },
          ticks: { color: COLOR_MUTED, font: { size: 10 }, backdropColor: 'transparent' },
        },
      },
    },
  };
}

// ── Main render function ─────────────────────────────────────────────────────
function render(cfg) {
  const {
    type = 'line',
    title = '',
    subtitle = '',
    source = '',
    labels = [],
    datasets = [],
    width = 900,
    height = 480,
    yMin: ymin = null,
    yMax: ymax = null,
    yPrefix: pre = '',
    ySuffix: suf = '',
  } = cfg;

  const BUILDERS = {
    line: cfgLine, bar: cfgBar, hbar: cfgHBar, area: cfgArea, scatter: cfgScatter,
    donut: cfgDonut, pie: cfgPie, polarArea: cfgPolarArea, radar: cfgRadar,
  };
  const { type: chartType, data, typeOptions } = (BUILDERS[type] ?? BUILDERS.line)({
    labels, datasets, pre, suf, ymin, ymax,
  });

  const hasHeader = !!(title || subtitle);
  const topPad = hasHeader ? (title && subtitle ? 56 : 38) : 12;
  const bottomPad = source ? 20 : 6;
  const isCircular = ['donut', 'pie', 'polarArea'].includes(type);
  const multiSeries = datasets.length > 1;

  const legendCfg = isCircular
    ? {
      display: true,
      position: 'right',
      labels: {
        color: COLOR_MUTED,
        font: { size: 10 },
        padding: 12,
        boxWidth: 12,
        boxHeight: 12,
        generateLabels: chart =>
          chart.data.labels.map((lbl, i) => ({
            text: `${lbl}  ${chart.data.datasets[0].data[i]}`,
            fillStyle: chart.data.datasets[0].backgroundColor[i],
            strokeStyle: '#ffffff',
            lineWidth: 1,
            hidden: false,
            index: i,
          })),
      },
    }
    : multiSeries
      ? {
        display: true,
        position: 'top',
        align: 'end',
        labels: { color: COLOR_MUTED, font: { size: 10 }, padding: 8, boxWidth: 10, boxHeight: 10 },
      }
      : { display: false };

  // Set up canvas with DOM-like properties Chart.js may touch
  const canvas = createCanvas(width, height);
  canvas.style = {};
  canvas.addEventListener = () => { };
  canvas.removeEventListener = () => { };
  canvas.getAttribute = (attr) => {
    if (attr === 'width') return canvas.width;
    if (attr === 'height') return canvas.height;
    return null;
  };
  canvas.setAttribute = (attr, value) => {
    if (attr === 'width') canvas.width = Number(value);
    if (attr === 'height') canvas.height = Number(value);
  };

  const ctx = canvas.getContext('2d');

  new Chart(ctx, {
    type: chartType,
    data,
    options: {
      responsive: false,
      animation: false,
      devicePixelRatio: 1,
      ...typeOptions,
      layout: {
        padding: { top: topPad, bottom: bottomPad, left: 8, right: 8 },
      },
      plugins: {
        legend: legendCfg,
        headerFooter: { title, subtitle, source },
      },
    },
  });

  return canvas.toBuffer('image/png');
}

// ── CLI entry point ──────────────────────────────────────────────────────────
const [, , arg1, outPath] = process.argv;

const HELP_TEXT = `
Beautiful Chart Renderer — Generate PNG charts from JSON config

USAGE
  bunx -y beautiful-chartsjs <config.json> [output.png]
  bunx -y beautiful-chartsjs --help | -h

ARGUMENTS
  config.json   Path to JSON configuration file (required)
  output.png    Output PNG file path (optional)
                - If provided: saves PNG to file
                - If omitted: prints base64 JSON to stdout

CHART TYPES
  line        Line chart with points
  bar         Vertical bar chart
  hbar        Horizontal bar chart
  area        Filled area chart
  scatter     Bubble/scatter chart
  donut       Donut chart
  pie         Pie chart
  polarArea   Polar area chart
  radar       Radar chart

CONFIG OPTIONS
  type        Chart type (default: "line")
  title       Chart title
  subtitle    Chart subtitle
  source      Source attribution (shown at bottom)
  labels      X-axis labels (array)
  datasets    Array of { label, data, color, fill? }
  width       Canvas width in px (default: 900)
  height      Canvas height in px (default: 480)
  yMin        Y-axis minimum value
  yMax        Y-axis maximum value
  yPrefix     Y-axis value prefix (e.g., "$")
  ySuffix     Y-axis value suffix (e.g., "%")

COLORS
  blue, red, teal, amber, purple, gray

EXAMPLE
  bunx -y beautiful-chartsjs chart.json output.png
`;

if (arg1 === '--help' || arg1 === '-h') {
  console.log(HELP_TEXT);
  process.exit(0);
}

if (!arg1) {
  console.error('Usage: bunx -y beautiful-chartsjs config.json [output.png]');
  console.error('Run with --help for more information.');
  process.exit(1);
}

const cfgPath = arg1;

let cfg;
try {
  cfg = JSON.parse(readFileSync(cfgPath, 'utf8'));
} catch (e) {
  console.error(`Error reading config: ${e.message}`);
  process.exit(1);
}

const png = render(cfg);

if (outPath) {
  writeFileSync(outPath, png);
  console.error(`saved → ${outPath}`);
} else {
  process.stdout.write(
    JSON.stringify({
      type: 'image',
      source: { type: 'base64', media_type: 'image/png', data: png.toString('base64') },
    }) + '\n',
  );
}
